extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java

```diff
@@ -0,0 +1,190 @@
+                        deferredThrowable = e;
+                    }
+                }
+            }
+
+            status = ExecutionStatus.FINISHED;
+        }
+
+        @Override
+        public void afterCompletion(int status) {
+            // The list should be iterated in reverse order
+            for (int i = synchs.size(); i-- > 0;) {
+                synchs.get(i).afterCompletion(status);
+            }
+        }
+
+        // does packageName belong to this group of synchronizations
+        private boolean shouldAdd(String packageName) {
+            return !packagePrefix.isEmpty() && packageName.startsWith(packagePrefix);
+        }
+    }
+
+    public AgroalOrderedLastSynchronizationList(
+            TransactionSynchronizationRegistryWrapper transactionSynchronizationRegistryWrapper) {
+
+        this.tsr = transactionSynchronizationRegistryWrapper;
+
+        for (var packagePrefix : PKG_PREFIXES) {
+            var synchronizationGroup = new SynchronizationGroup(packagePrefix);
+
+            synchGroups.add(synchronizationGroup);
+
+            if (packagePrefix.isEmpty()) {
+                otherSynchs = synchronizationGroup; // the catch-all group
+            }
+        }
+    }
+
+    /**
+     * Register an interposed synchronization. Note that synchronizations are not allowed if:
+     * <p>
+     *
+     * @param synchronization The synchronization to register
+     * @throws IllegalStateException if the transaction is in the wrong state:
+     *         <ol>
+     *         <li>the transaction has already prepared;
+     *         <li>the transaction is marked rollback only
+     *         <li>the group that the synchronization should belong to has already been processed
+     *         </ol>
+     */
+    public void registerInterposedSynchronization(Synchronization synchronization) {
+        int status = tsr.getTransactionStatus();
+
+        switch (status) {
+            case Status.STATUS_ACTIVE:
+            case Status.STATUS_PREPARING:
+                break;
+            default:
+                throw new IllegalStateException(REGISTER_SYNC_ERROR + status);
+        }
+
+        // add the synchronization to the group that matches this package and, if there is no such group
+        // then add it to the catch-all group (otherSyncs)
+        String packageName = synchronization.getClass().getName();
+        SynchronizationGroup synchGroup = otherSynchs;
+
+        for (SynchronizationGroup g : synchGroups) {
+            if (g.shouldAdd(packageName)) {
+                synchGroup = g;
+                break;
+            }
+        }
+
+        synchGroup.add(synchronization);
+    }
+
+    /**
+     * Exceptions from beforeCompletion Synchronizations are not caught because such errors should cause the
+     * transaction to roll back.
+     */
+    @Override
+    public void beforeCompletion() {
+        // run each group of synchs according to the order they were added to the list
```

Comment:
I expect that both beforeCompletion + afterCompletion method calls will invoke `Hibernate` second to last and `Agroal` last.  I'm not really sure though if the above code is always adding `Hibernate` second to last in the list and `Agroal` last in the list.  Did I miss something that ensures that Hibernate/Agroal are added to the list last?

Reply:
The synchronization groups are maintained in an ordered list [1]. There are three such groups and the members are based on the package names [2] and [3] of the synchronization being registered. The order of the package names [3] is '{ "", "org.hibernate", "io.agroal.narayana" };` so the second group of synchronizations is any synchronization with package prefix "org.hibernate" and the last group contains synchronizations with package prefix "io.agroal.narayana".

The code you have highlighted iterates through each group in order and calls `beforeCompletion()` on the group which in turn calls `beforeCompletion()` on the actual registered synchronizations [5].

[1] https://github.com/mmusgrov/quarkus/blob/18737-synch-order/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java#L21
[2] https://github.com/mmusgrov/quarkus/blob/18737-synch-order/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java#L115
[3] https://github.com/mmusgrov/quarkus/blob/18737-synch-order/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java#L19
[4] https://github.com/mmusgrov/quarkus/blob/18737-synch-order/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java#L168
[5] https://github.com/mmusgrov/quarkus/blob/18737-synch-order/extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/internal/tsr/AgroalOrderedLastSynchronizationList.java#L71